File: android/go/src/main/java/com/phpnitro/go/HostPort.kt
package com.phpnitro.go

import android.content.Context
import android.content.Intent

/** Both ConnectActivity's manual "Connecter" and ScanActivity's decoded QR end up here — same target, same extras. */
fun renderIntent(context: Context, host: String, port: Int): Intent = Intent().apply {
    setClassName(context, "com.phpnitro.engine.NativeRenderPocActivity")
    putExtra("serverHost", host)
    putExtra("serverPort", port)
    putExtra("screen", "home")
}

/**
 * Shared by ConnectActivity's manual entry field and ScanActivity's decoded
 * QR payload — both ultimately need the same "IP:PORT, optionally prefixed
 * with a scheme" parsing, since `phpx serve`'s printed URL and its QR
 * code encode the exact same string (see bin/QrCode.php's caller in
 * cmdServe()).
 */
fun parseHostPort(input: String): Pair<String, Int>? {
    val withoutScheme = input.trim().removePrefix("http://").removePrefix("https://").trimEnd('/')
    val colonIndex = withoutScheme.lastIndexOf(':')
    if (colonIndex <= 0 || colonIndex == withoutScheme.length - 1) return null
    val host = withoutScheme.substring(0, colonIndex)
    val port = withoutScheme.substring(colonIndex + 1).toIntOrNull() ?: return null
    if (host.isEmpty() || port !in 1..65535) return null
    return host to port
}

File: ios/Sources/PhpNitroGo/HostPort.swift
import Foundation

/// The iOS counterpart of android/go's HostPort.kt — shared by
/// ConnectViewController's manual entry field and (once it exists) an
/// iOS QR scanner, since `phpx serve`'s printed URL and its QR code
/// encode the exact same "IP:PORT, optionally prefixed with a scheme"
/// string on both platforms.
enum HostPort {
    static func parse(_ input: String) -> (host: String, port: Int)? {
        var withoutScheme = input.trimmingCharacters(in: .whitespacesAndNewlines)
        for prefix in ["http://", "https://"] where withoutScheme.hasPrefix(prefix) {
            withoutScheme.removeFirst(prefix.count)
            break
        }
        while withoutScheme.hasSuffix("/") {
            withoutScheme.removeLast()
        }

        guard let colonIndex = withoutScheme.lastIndex(of: ":"), colonIndex != withoutScheme.startIndex else { return nil }

        let host = String(withoutScheme[withoutScheme.startIndex..<colonIndex])
        let portString = String(withoutScheme[withoutScheme.index(after: colonIndex)...])

        guard !host.isEmpty, !portString.isEmpty, let port = Int(portString), (1...65535).contains(port) else { return nil }

        return (host, port)
    }
}
